python/cudf/cudf/core/indexed_frame.py

```diff
@@ -526,3 +532,208 @@ def _n_largest_or_smallest(self, largest, n, columns, keep):
             return self.take(indices, keep_index=True)
         else:
             raise ValueError('keep must be either "first", "last"')
+
+    def _align_to_index(
+        self: T,
+        index: ColumnLike,
+        how: str = "outer",
+        sort: bool = True,
+        allow_non_unique: bool = False,
+    ) -> T:
+        index = cudf.core.index.as_index(index)
+
+        if self.index.equals(index):
+            return self
+        if not allow_non_unique:
+            if len(self) != len(self.index.unique()) or len(index) != len(
+                index.unique()
+            ):
+                raise ValueError("Cannot align indices with non-unique values")
+
+        lhs = cudf.DataFrame._from_data(self._data, index=self.index)
+        rhs = cudf.DataFrame(index=index)
+
+        tmp_col_id = str(uuid4())
+        if how == "left":
+            lhs[tmp_col_id] = arange(len(lhs))
+        elif how == "right":
+            rhs[tmp_col_id] = arange(len(rhs))
+
+        result = lhs.join(rhs, how=how, sort=sort)
+        if how == "left" or how == "right":
+            result = result.sort_values(tmp_col_id)
+            del result[tmp_col_id]
+
+        result = self.__class__._from_data(result._data, index=result.index)
```

Comment:
Just to make sure I understand, for a 1D frame (Series) this function is going to make a 2-col DataFrame for the purposes of sort order but then just extract the index-sorted original column back out to reconstruct, right? I don't really like that we have to refer to DataFrame in this class, but I don't know if there's a great alternative since moving the impl of some of these methods to Frame would expose Series methods that shouldn't exist.

This is probably out of scope for this PR, these types of logical circular references exist in multiple places in cudf, but it's something I'd like to address eventually. We should think a bit about whether we can/should move all Series methods that require conversion to DataFrame to the Series level rather than in IndexedFrame. Otherwise it's bound to lead to inconsistent implementations at worst, and less efficient implementations at best.

Reply:
Correct -- I don't mind so much referring to subclasses in a base class. It's the reason we like dynamic languages like Python. I _don't_ like that we have to construct a DataFrame for this operation, where ideally some simpler, internal frame-like data structure would suffice.

Reply:
IMO while it's the kind of thing that dynamic languages _allow_, it is usually (but of course not always) a code smell. It's not that different from forward declarations or friends in C++; there's a time and place, but often it often indicates that there's a better design to be had. But that's something we can debate at a later point when we're not coming up on code freeze :)